### PR TITLE
Terraform additions

### PR DESCRIPTION
**Pull request by eitan-kr**
Asm 12282 Fix issue in akeyless pki cert issuer
Asm 12364 Add k8s auth client cert params and use gw sa
Asm 11955 Add classic key support in TF provider

### CODE REVIEW

File: akeyless/common/utils.go [hunk 1]
```diff
@@ -547,3 +547,60 @@ func ExtractLogForwardingFormat(isJson bool) string {
...
+		expirationEventsList = append(expirationEventsList, strconv.FormatInt(days, 10))
+	}
+	return expirationEventsList
+}
+
+func UpdateRotationSettings(d *schema.ResourceData, name string, token string, client akeyless_api.V2ApiService) error {
+	var apiErr akeyless_api.GenericOpenAPIError
+	ctx := context.Background()
+
+	if d.HasChanges("auto_rotate", "rotation_interval", "rotation_event_in") {
```
Comment (tuvia-akeyless) [inline akeyless/common/utils.go:575]: What will happen if only few of them are set?
- if `rotation_interval` not set - the Atoi may fail.
- if `auto_rotate` not set - the ParseBool may fail.
- if `rotation_event_in` not set - what will happen when building list? would it be compatible with the Read step results?
Maybe all these issues are the expected behavior, but anyway it should be considered.

File: akeyless/resource_dfc_key.go [hunk 2]
```diff
@@ -427,6 +513,7 @@ func validateDfcKeyUpdateParams(d *schema.ResourceData) error {
 	paramsMustNotUpdate := []string{"alg", "split_level", "customer_frg_id",
 		"generate_self_signed_certificate", "certificate_ttl",
 		"certificate_common_name", "certificate_organization",
-		"certificate_country", "certificate_locality", "certificate_province"}
+		"certificate_country", "certificate_locality", "certificate_province",
+		"conf_file_data", "certificate_format"}
```
Comment (tuvia-akeyless) [inline akeyless/resource_dfc_key.go:517]: `certificate_format` will be supported for update in next gateway. If gateway will release before TF, please add this support here as well.

File: akeyless/common/utils.go [hunk 3]
```diff
@@ -547,3 +547,60 @@ func ExtractLogForwardingFormat(isJson bool) string {
...
+	var expirationEventsList []string
+	for _, e := range expirationEvents {
+		seconds := e.GetSecondsBefore()
+		days := seconds / 60 / 60 / 24
+		expirationEventsList = append(expirationEventsList, strconv.FormatInt(days, 10))
+	}
+	return expirationEventsList
+}
+
+func UpdateRotationSettings(d *schema.ResourceData, name string, token string, client akeyless_api.V2ApiService) error {
```
Comment (tuvia-akeyless) [inline akeyless/common/utils.go:571]: I prefer to put this func in the resource code, and not in common. In addition, I prefer to take the `if` outside this func (call the func inside the `if`)

File: akeyless/resource_gateway_create_k8s_auth_config.go [hunk 4]
```diff
@@ -245,6 +274,30 @@ func resourceK8sAuthConfigRead(d *schema.ResourceData, m interface{}) error {
 			return err
 		}
 	}
+	if rOut.UseLocalCaJwt != nil {
+		err = d.Set("use_local_ca_jwt", *rOut.UseLocalCaJwt)
```
Comment (tuvia-akeyless) [inline akeyless/resource_gateway_create_k8s_auth_config.go:278]: Please add test for it and ensure it pass locally. Or at least check it manually to ensure you don't need any additions in the read results here.

File: akeyless/resource_item_test.go [hunk 5]
```diff
@@ -54,6 +56,8 @@ func TestDfcKeyRsaResource(t *testing.T) {
 			certificate_province 				= "prov1"
 			cert_data_base64 					= "%v"
 			tags        						= ["t1","t3"]
+			certificate_format = "der"
```
Comment (tuvia-akeyless) [inline akeyless/resource_item_test.go:59]: Please add test case for pem. If gateway released, please check case of update from der to pem and back

File: akeyless/resource_classic_key.go [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Description: "Whether to generate a self signed certificate with the key. If set, certificate_ttl must be provided.",
+				Default:     "false",
+			},
+			"certificate_ttl": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Description: "TTL in days for the generated certificate. Required only for generate-self-signed-certificate.",
+				Default:     0,
```
Comment (tuvia-akeyless) [inline akeyless/resource_classic_key.go:82]: 0 can't be real default value. If you got issues with it, please solve them in the read func.

File: akeyless/resource_classic_key.go [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+	common.GetAkeylessPtr(&body.RotationInterval, rotationInterval)
+	common.GetAkeylessPtr(&body.RotationEventIn, rotationEventIn)
+	common.GetAkeylessPtr(&body.DeleteProtection, deleteProtection)
+
+	_, _, err := client.CreateClassicKey(ctx).Body(body).Execute()
+	if err != nil {
+		if errors.As(err, &apiErr) {
+			return fmt.Errorf("can't create Secret: %v", string(apiErr.Body()))
+		}
+		return fmt.Errorf("can't create Secret: %v", err)
```
Comment (tuvia-akeyless) [inline akeyless/resource_classic_key.go:218]: secret --> classic-key (in all error msgs in this resource)

File: akeyless/resource_classic_key.go [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+	}
+	return keyType, nil
+}
+
+func validateClassicKeyUpdateParams(d *schema.ResourceData) error {
+	paramsMustNotUpdate := []string{"alg", "gpg_alg",
+		"generate_self_signed_certificate", "certificate_ttl",
+		"certificate_common_name", "certificate_organization",
+		"certificate_country", "certificate_locality", "certificate_province",
+		"conf_file_data", "certificate_format"}
```
Comment (tuvia-akeyless) [inline akeyless/resource_classic_key.go:486]: `certificate_format` would be valid here too if gateway released.

File: akeyless/resource_item_test.go [hunk 9]
```diff
@@ -78,6 +87,11 @@ func TestDfcKeyAesResource(t *testing.T) {
 			name = "%v"	
 			tags     = ["t1", "t3"]
 			alg = "AES128SIV"
+			certificate_format = "der"
+			auto_rotate = "true"
+			rotation_interval = "9"
+			rotation_event_in = ["4"]
```
Comment (tuvia-akeyless) [inline akeyless/resource_item_test.go:93]: try also update with: `auto_rotate = "false"` and with empty `rotation_interval` & `rotation_event_in`

File: akeyless/resource_item_test.go [hunk 10]
```diff
@@ -117,16 +131,68 @@ func TestClassicKey(t *testing.T) {
...
 		resource "akeyless_classic_key" "%v" {
 			name 		= "%v"	
 			alg 		= "RSA2048"
+			generate_self_signed_certificate 	= true
+			certificate_ttl 					= 60
+			certificate_common_name 			= "cn1"
+			certificate_organization 			= "org1"
+			certificate_country 				= "cntry1"
+			certificate_locality 				= "local1"
+			certificate_province 				= "prov1"
```
Comment (tuvia-akeyless) [inline akeyless/resource_item_test.go:160]: try to update values of all these params...

File: akeyless/resource_item_test.go [hunk 11]
```diff
@@ -117,16 +131,68 @@ func TestClassicKey(t *testing.T) {
...
+			expiration_event_in = ["14"]
+		}
+	`, name, itemPath)
+
+	testItemResource(t, itemPath, config, configUpdate)
+}
+
+func TestClassicGpgKey(t *testing.T) {
+
+	t.Skip("not authorized to create producer on public gateway")
```
Comment (tuvia-akeyless) [inline akeyless/resource_item_test.go:176]: it isn't producer...

File: akeyless/resource_item_test.go [hunk 12]
```diff
@@ -117,16 +131,68 @@ func TestClassicKey(t *testing.T) {
...
+			alg			= "GPG"
+			gpg_alg 	= "RSA2048"
+		}
+	`, name, itemPath)
+
+	configUpdate := fmt.Sprintf(`
+		resource "akeyless_classic_key" "%v" {
+			name 		= "%v"	
+			alg			= "GPG"
+			gpg_alg 	= "RSA2048"
```
Comment (tuvia-akeyless) [inline akeyless/resource_item_test.go:194]: I think `gpg_alg` isn't updatable. at list via UI. Please check if it should be added to the not-update list...

Comment (eitan-kr) [reply akeyless/resource_item_test.go:194]: Already in list

File: akeyless/resource_classic_key.go [hunk 13]
```diff
@@ -0,0 +1,488 @@
...
+	if err == nil {
+		if len(add) > 0 {
+			common.GetAkeylessPtr(&body.AddTag, add)
+		}
+		if len(remove) > 0 {
+			common.GetAkeylessPtr(&body.RmTag, remove)
+		}
+	}
+
+	_, _, err = client.UpdateItem(ctx).Body(body).Execute()
```
Comment (tuvia-akeyless) [inline akeyless/resource_classic_key.go:416]: Use GatewayUpdateItem

File: akeyless/resource_gateway_create_k8s_auth_config.go (hunk 4 above)
Comment (eitan-kr) [reply akeyless/resource_gateway_create_k8s_auth_config.go:278]: For this we need the GW to run in k8s...

Comment (tuvia-akeyless) [reply akeyless/resource_gateway_create_k8s_auth_config.go:278]: Ok, can you check it manually?

File: akeyless/resource_classic_key.go (hunk 13 above)
Comment (eitan-kr) [reply akeyless/resource_classic_key.go:415]: But for that we have UpdateRotationSettings

**Review by michelsk — APPROVED** [review]

**Review by tuvia-akeyless — APPROVED** [review]

**Pull request merged**